### ISSUE TITLE
Wrong tilt calculation for the aqara vibration sensor

### ISSUE THREAD

**Issue by mariusGundersen** [TedTolboom/com.xiaomi-mi-zigbee#319]
I'm fairly certain the [code](https://github.com/TedTolboom/com.xiaomi-mi-zigbee/blob/master/drivers/vibration.aq1/device.js#L146)  for finding the tilt angle is incorrect. It should probably use dot product between the incoming vector and a stored vector to find the tilt angle and also the relative tilt angle. As it is now it completely ignores one axis when finding the tilt angle, which means it only works when the device is positioned and tilted in certain ways.

If this repository is still alive and maintained then I'll try to create a pull request next week

**Comment by TedTolboom** [comment]
Yes it is still maintained.I’m looking forward to your contributionOp 6 apr. 2023 om 18:12 heeft Marius Gundersen ***@***.***> het volgende geschreven:﻿
I'm fairly certain the code  for finding the tilt angle is incorrect. It should probably use dot product between the incoming vector and a stored vector to find the tilt angle and also the relative tilt angle. As it is now it completely ignores one axis when finding the tilt angle, which means it only works when the device is positioned and tilted in certain ways.
If this repository is still alive and maintained then I'll try to create a pull request next week

**Comment by mariusGundersen** [comment]
Ok, I've been looking into this a bit, and first of all [the way the raw buffer is decoded](https://github.com/TedTolboom/com.xiaomi-mi-zigbee/blob/master/drivers/vibration.aq1/device.js#L137) doesn't work.

```js
    // FIX data as parsed by the Zigbee Shepherd
    const toInt16 = v => Int16Array.from([v])[0];

    const Rx = toInt16((value[1]) & 0xffff); toInt16((value[1]) & 0xffff);
    const Ry = toInt16((value[1] >> 16) & 0xffff);
    const Rz = toInt16(value[0]);
```

since `value` is a buffer each entry contains an 8 bit number, so shifting it down by 16 bits will return 0. So `Ry` is always 0. `value` is a 6 byte buffer, which can be interpreted as 3 16 bit integers if done like this:

```js
    const [Rx, Ry, Rz] = new Int16Array(value.buffer, value.byteOffset, value.byteLength);
```

Using this I got a lot better results

**Comment by mariusGundersen** [comment]
The results I got were 16 bit values between roughly -1000 and 1000, which seems natural if it measures 1g using 10 bits, which would be 1024. But I did get some strange results for the z axis, which seems to report between -600 and 1400, which meant that when I put the tilt sensor on the side it ended up at 72 rather than 90 degrees. For fun I tried subtracting 350 from the z axis value and that gave me much cleaner 90 degree values. I'm not sure if this is a consistent thing or not, I have two of these vibration sensors and both of them are equally offset on the z axis. Maybe the z axis offset could be a setting that could be adjusted?
